src/core/geometry/qgsgeometry.cpp

```diff
@@ -1447,6 +1447,28 @@ double QgsGeometry::distance( const QgsGeometry &geom ) const
   return g.distance( geom.d->geometry );
 }
 
+double QgsGeometry::hausdorffDistance( const QgsGeometry &geom ) const
+{
+  if ( !d->geometry || !geom.d->geometry )
+  {
+    return -1.0;
+  }
+
+  QgsGeos g( d->geometry );
+  return g.hausdorffDistance( geom.d->geometry );
```

Comment:
What about a bit of error reporting?

```cpp
return g.hausdorffDistance( geom.d->geometry, d->error );
```

Reply:
Wouldn't that require a detach?

Reply:
Right.
What do you think about switching the `QString* errorMsg = nullptr` param to a `std::function < void( const QString& error ) > errorCallback = nullptr`?

Reply:
Hmm. Sounds a bit complex to use. Maybe we should just SIP_OUT a string pointer and leave it at that?

Reply:
I think the SIP_OUT approach sounds complex to use, it will require creating a new QString and passing it into QgsGeometry functions every single time you call them in C++. I'd prefer to have all this handling hidden away in QgsGeometry, that's why I introduced `QgsGeometry::error()` which should either be used always or never.

Reply:
Alternative idea: we move the error string out of the private geometry class and it becomes a mutable member just of QgsGeometry, not copied when QgsGeometry is copied.

Reply:
Is there any risk involved with this combined with multithreading?

Reply:
Forget that. Better question.

Will this work with

```cpp
QgsGeometry QgsGeometry::buffer( double distance )
{
   QgsGeometry result;
   result.mError = tr( "Something has gone wrong" );
   return result;
}
```

Reply:
Ok, it'll need to be copied too.